Wrappers/Python/ccpi/viewer/utils/hdf5_io.py

```diff
@@ -0,0 +1,184 @@
+import vtk
+from vtk.util.vtkAlgorithm import VTKPythonAlgorithmBase
+import h5py
+import numpy as np
+from vtk.numpy_interface import dataset_adapter as dsa
+
+# Methods for reading and writing HDF5 files:
+
+
+def write_image_data_to_hdf5(filename, data, label="ImageData", attributes={},
+                             array_name='vtkarray'):
+    '''
+    Writes vtkImageData to a dataset in a HDF5 file
+
+    Args:
+        filename: string - name of HDF5 file to write to.
+        data: vtkImageData - image data to write.
+        label: string - label for HDF5 dataset.
+        attributes: dict - attributes to assign to HDF5 dataset.
+        array_name: string - name of array to read points from the vtkImageData
+    '''
+
+    with h5py.File(filename, "a") as f:
+        wdata = dsa.WrapDataObject(data)
+        array = wdata.PointData[array_name]
+        # Note that we flip the dimensions here because
+        # VTK's order is Fortran whereas h5py writes in
+        # C order. We don't want to do deep copies so we write
+        # with dimensions flipped and pretend the array is
+        # C order.
+        array = array.reshape(wdata.GetDimensions()[::-1])
+        try:
+            dset = f.create_dataset(label, data=array)
+        except RuntimeError:
+            print("Unable to save image data to {0}."
+                  "Dataset with name {1} already exists in this file.".format(
+                      filename, label))
+            return
+        for key, value in attributes.items():
+            dset.attrs[key] = value
+        # print("The file written: ", f)
+        # print("The image data: ", f[label])
+        # print("The attributes: ", f[label].attrs.items())
+
+
+class HDF5Reader(VTKPythonAlgorithmBase):
+    '''
+    vtkAlgorithm for reading vtkImageData from a HDF5 file
+    Adapted from:
+    https://blog.kitware.com/developing-hdf5-readers-using-vtkpythonalgorithm/
+    '''
+
+    def __init__(self):
+        VTKPythonAlgorithmBase.__init__(self,
+                                        nInputPorts=0,
+                                        nOutputPorts=1,
+                                        outputType='vtkImageData')
+
+        self.__FileName = ""
+        self.__Label = "ImageData"
```

Comment:
is this reader reading NeXus as exported by CIL only?

Reply:
No, it's not, you have to set the label where the data lives, maybe it's best that I change the default to:
`'entry1/tomo_entry/data/data'`

Reply:
set to None by default and renamed to self.__DatasetName